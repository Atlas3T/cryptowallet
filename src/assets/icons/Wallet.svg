<?xml version="1.0" encoding="utf-8"?>
<!-- Generator: Adobe Illustrator 22.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
<svg version="1.1" id="Layer_1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" x="0px" y="0px"
	 viewBox="0 0 51 51" style="enable-background:new 0 0 51 51;" xml:space="preserve">
<style type="text/css">
	.st0{fill:none;stroke:#000000;stroke-width:5;stroke-miterlimit:10;}
	.st1{display:none;}
	.st2{display:inline;fill:none;stroke:#000000;stroke-linecap:round;stroke-linejoin:round;stroke-miterlimit:10;}
	.st3{display:inline;}
</style>
<g id="Wallet">
	<path class="st0" d="M46.1,31.5v5.3c0,2.4-1.9,4.3-4.3,4.3H9.5c-2.4,0-4.3-1.9-4.3-4.3V15.5c0-2.4,1.9-4.3,4.3-4.3h32.3
		c2.4,0,4.3,1.9,4.3,4.3v5.2"/>
	<circle cx="46.1" cy="26.1" r="3.1"/>
</g>
<g id="Payments" class="st1">
	<line class="st2" x1="8" y1="19.8" x2="41.8" y2="19.8"/>
	<polyline class="st2" points="35.8,13.3 42.3,19.8 35.8,26.3 	"/>
	<line class="st2" x1="42.3" y1="32.8" x2="8.5" y2="32.8"/>
	<polyline class="st2" points="14.5,39.3 8,32.8 14.5,26.3 	"/>
</g>
<g id="Exchange" class="st1">
	<path class="st2" d="M15.3,21.7c0-5.6,4.6-10.2,10.2-10.2s10.2,4.6,10.2,10.2"/>
	<polyline class="st2" points="42.2,16.3 35.7,22.8 29.2,16.3 	"/>
	<path class="st2" d="M35.7,30.2c0,5.6-4.6,10.2-10.2,10.2s-10.2-4.6-10.2-10.2"/>
	<polyline class="st2" points="8.8,35.7 15.3,29.2 21.9,35.7 	"/>
</g>
<g id="More" class="st1">
	<circle class="st3" cx="26" cy="15.8" r="7.8"/>
	<circle class="st3" cx="12.3" cy="28.6" r="5"/>
	<circle class="st3" cx="35.3" cy="33.6" r="3.5"/>
</g>
</svg>
