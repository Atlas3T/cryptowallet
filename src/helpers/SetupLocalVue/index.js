import Vuex from 'vuex';
import { createLocalVue } from '@vue/test-utils';
import VueRouter from 'vue-router';
import VueI18n from 'vue-i18n';
import messages from '@/i18n/';
import axios from 'axios';
import { sync } from 'vuex-router-sync';
import errorHandlerPlugin from '@/boot/ErrorHandler';
import toasterPlugin from '@/boot/Toaster';
import permissionsPlugin from '@/boot/Permissions';
import appInvitationPlugin from '@/boot/AppInvitation';
import smsPlugin from '@/boot/Sms';
import emailPlugin from '@/boot/Email';
import contactsImportPlugin from '@/boot/ContactsImport';
import * as All from 'quasar'

const { Quasar, date } = All;

const components = Object.keys(All).reduce((object, key) => {
  const val = All[key]
  if (val && val.component && val.component.name != null) {
    object[key] = val
  }
  return object
}, {})


const localVue = createLocalVue();

localVue.use(Quasar, { components });
localVue.use(VueI18n);
localVue.use(Vuex);
// localVue.use(VueRouter);
localVue.prototype.$axios = axios;
errorHandlerPlugin({ Vue: localVue });
toasterPlugin({ Vue: localVue });
permissionsPlugin({ Vue: localVue });
appInvitationPlugin({ Vue: localVue });
smsPlugin({ Vue: localVue });
emailPlugin({ Vue: localVue });
contactsImportPlugin({ Vue: localVue });

/**
 * creates a new router and syncs it with the store using vuex-router-sync
 * @param store
 * @returns router
 */
const createRouter = (store = null) => {
  const router = new VueRouter({
    routes: [
      {
        path: '/contact/:id',
      },
    ],
  });

  if (store) { sync(store, router); }
  return router;
};

const i18n = new VueI18n({ messages, locale: 'en-gb' });

export {
  localVue,
  createRouter,
  i18n,
};
