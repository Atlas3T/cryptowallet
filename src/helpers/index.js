import AmountFormatter from './AmountFormatter';
import dateTranslater from './DateTranslater';
import Contacts from './Contacts';

export {
  AmountFormatter,
  dateTranslater,
  Contacts,
};
