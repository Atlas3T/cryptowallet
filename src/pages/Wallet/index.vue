<template>
  <div class="layout">
    <div class="list">
      <wallet-list
        :wallets="wallets"
        :currency="selectedCurrency"
      />
    </div>
    <div class="button">
      <add-wallet/>
    </div>
  </div>
</template>

<script>
import WalletList from '@/components/Wallet/WalletsList';
import AddWallet from '@/components/Wallet/AddWallet';

export default {
  name: 'Balance',

  components: {
    WalletList,
    AddWallet,
  },

  /**
   * @TODO James Konrad
   * data below is mocked
   *
   * currency String is a prefered user currency, should be in the Vuex store
   * you should probably select your prefered currency during the initial setup
   * alongside the language
   *
   * coins Object also shoudn't be here
   */
  data() {
    return {
      wallets: {
        BTC: {
          key: 'BTC',
          name: 'Bitcoin',
          balance: 25.532344,
          uid: 1,
        },
        ATL: {
          key: 'ATL',
          name: 'Atlas',
          balance: 21.532344,
          uid: 2,
        },
        ETH: {
          key: 'ETH',
          name: 'Ethereum',
          balance: 55.532344,
          uid: 3,
        },
      },
    };
  },

  computed: {
    selectedCurrency() {
      return this.$store.state.settings.selectedCurrency;
    },
  },
};

</script>

<style scoped>
.layout {
  display: flex;
  flex-direction: column;
}

.list {
  flex: 1;
  position: relative;
}
</style>
