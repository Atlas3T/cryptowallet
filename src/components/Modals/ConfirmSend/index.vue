<template>
  <div>
    <q-dialog
      v-model="confirmSendModalOpened"
      persistent
      :maximized="true"
      transition-show="slide-up"
      transition-hide="slide-down"
      content-class="light-modal"
    >
      <div
        :class="{ active: loading }"
        class="sending-spinner-overlay"
      >
        <Spinner />
      </div>

      <div class="header-section">
        <div class="header-back-button-wrapper">
          <q-btn
            icon="arrow_back"
            size="lg"
            class="icon-btn back-arrow-btn"
            flat
            @click.prevent="goBack"
          />
        </div>
        <h1 class="header-h1">
          {{ $t('confirm') }}
        </h1>
      </div>
      <div
        v-if="wallet && txData"
        class="modal-layout-wrapper"
      >
        <CoinHeader :wallet="wallet" />

        <div class="send-modal-heading">
          <h3>{{ $t('to') }}</h3>
          <span class="h3-line" />
        </div>

        <div class="small-text break">
          {{ to }}
        </div>

        <div class="send-modal-heading">
          <h3>{{ $t('amount') }}</h3>
          <span class="h3-line" />
        </div>

        <div class="small-text">
          {{ txData.transaction.value }} {{ coinSymbol }}
          ({{ coinToCurrency(txData.transaction.value) }})
        </div>

        <div class="send-modal-heading">
          <h3>{{ $t('fee') }}</h3>
          <span class="h3-line" />
        </div>

        <div
          v-if="!isErc20"
          class="small-text"
        >
          {{ txData.transaction.fee }} {{ coinSymbol }}
          ({{ coinToCurrency(txData.transaction.fee, true) }})
        </div>
        <div
          v-if="isErc20"
          class="small-text"
        >
          {{ txData.transaction.fee / weiMultiplier }}  {{ $t('ethSymbol') }}
          ({{ coinToCurrency(txData.transaction.fee / weiMultiplier, true) }})
        </div>

        <div class="send-modal-heading">
          <h3>{{ $t('newBalance') }}</h3>
          <span class="h3-line" />
        </div>

        <div class="small-text">
          {{ newBalance }} {{ coinSymbol }}
          ({{ coinToCurrency(newBalance) }})
        </div>

        <div class="send">
          <q-btn
            :label="$t('Confirm')"
            color="blueish"
            size="md"
            @click="confirm"
          />
        </div>
      </div>
    </q-dialog>
  </div>
</template>

<script>
import { mapState } from 'vuex';
import CoinHeader from '@/components/Wallet/CoinHeader';
import Spinner from '@/components/Spinner';
import Address from '@/store/wallet/entities/address';
import Wallet from '@/store/wallet/entities/wallet';
import Tx from '@/store/wallet/entities/tx';
import Utxo from '@/store/wallet/entities/utxo';
import Coin from '@/store/wallet/entities/coin';
import {
  AmountFormatter,
  getBalance,
} from '@/helpers';

export default {
  name: 'ConfirmSend',
  components: {
    CoinHeader,
    Spinner,
  },

  data() {
    return {
      confirmSendModalOpened: false,
      txData: null,
      loading: false,
      weiMultiplier: 1000000000000000000,
    };
  },

  computed: {
    ...mapState({
      id: (state) => { return state.route.params.id; },
      authenticatedAccount: (state) => { return state.settings.authenticatedAccount; },
      delay: (state) => { return state.settings.delay; },
    }),

    wallet() {
      return this.$store.getters['entities/wallet/find'](this.id);
    },

    supportedCoins() {
      return Coin.all();
    },

    selectedCurrency() {
      return this.$store.state.settings.selectedCurrency;
    },

    coinSymbol() {
      return this.supportedCoins.find((coin) => { return coin.name === this.wallet.name; }).symbol;
    },

    newBalance() {
      if (this.wallet.sdk === 'Ethereum') {
        const newBalance = (this.wallet.confirmedBalance * this.weiMultiplier)
                           - ((this.txData.transaction.value * this.weiMultiplier)
                           + (parseFloat(this.txData.transaction.fee) * this.weiMultiplier));
        return newBalance / this.weiMultiplier;
      }
      if (this.wallet.sdk === 'ERC20') {
        const newBalance = (this.wallet.confirmedBalance * this.weiMultiplier)
                           - (this.txData.transaction.value * this.weiMultiplier);
        return newBalance / this.weiMultiplier;
      }
      if (this.wallet.sdk === 'Bitcoin') {
        const totalCost = this.txData.transaction.value + parseFloat(this.txData.transaction.fee);
        const unconfirmedBalance = getBalance(this.wallet, this.authenticatedAccount).unconfirmed;
        return unconfirmedBalance - totalCost;
      }
      return false;
    },

    to() {
      if (Array.isArray(this.txData.transaction.receiver)) {
        return this.txData.transaction.receiver[0];
      }
      return this.txData.transaction.receiver;
    },

    latestPrice() {
      const prices = this.$store.getters['entities/latestPrice/find'](`${this.coinSymbol}_${this.selectedCurrency.code}`);
      return prices.data.PRICE;
    },

    isErc20() {
      if (this.wallet.sdk === 'ERC20') {
        return true;
      }
      return false;
    },
  },

  mounted() {
    this.$root.$on('confirmSendModalOpened', (value, txData) => {
      this.confirmSendModalOpened = value;
      this.txData = txData;
    });
  },

  methods: {
    async broadcastTx() {
      const {
        hexTx,
        transaction,
        utxo,
        changeAddresses,
      } = this.txData;


      const coinSDK = this.coinSDKS[this.wallet.sdk];
      if (this.wallet.sdk === 'Bitcoin') {
        const result = await coinSDK.broadcastTx(hexTx, this.wallet.network);
        if (!result) {
          throw new Error(this.$t('transactionBroadcastFail'));
        }

        transaction.account_id = this.authenticatedAccount;
        transaction.wallet_id = this.wallet.id;
        transaction.isChange = false;
        transaction.sent = true;

        await Tx.$insert({ data: transaction });

        utxo.forEach((usedUtxo) => {
          const whereUtxo = (record, item) => {
            return (
              record.txid === item.txid
            && record.vout === item.vout
            && record.wallet_id === this.wallet.id
            );
          };

          Utxo.$update({
            where: (record) => {
              return whereUtxo(record, usedUtxo);
            },
            data: {
              pending: true,
            },
          });
        });

        changeAddresses.forEach(async (address, i) => {
          await Address.$insert({
            data: {
              address,
              account_id: this.authenticatedAccount,
              wallet_id: this.wallet.id,
              chain: 'internal',
              index: this.wallet.internalChainAddressIndex + i,
            },
          });
        });

        const newInternalIndex = this.wallet.internalChainAddressIndex + changeAddresses.length;

        Wallet.$update({
          where: (record) => { return record.id === this.wallet.id; },
          data: { internalChainAddressIndex: newInternalIndex },
        });

        this.completeTransaction();
        return false;
      }

      if (this.wallet.sdk === 'Ethereum' || this.wallet.sdk === 'ERC20') {
        const result = await coinSDK.broadcastTx(hexTx, this.wallet.network);
        if (!result) {
          throw new Error(this.$t('transactionBroadcastFail'));
        }

        transaction.account_id = this.authenticatedAccount;
        transaction.wallet_id = this.wallet.id;
        transaction.isChange = false;
        transaction.sent = true;

        await Tx.$insert({ data: transaction });

        this.completeTransaction();
        return false;
      }

      return false;
    },
    coinToCurrency(amount, fee) {
      let formattedAmount = new AmountFormatter({
        amount,
        rate: this.latestPrice,
        format: '0,0[.]00',
        coin: this.wallet.name,
        prependPlusOrMinus: false,
        currency: this.selectedCurrency,
        toCurrency: true,
        withCurrencySymbol: true,
      });

      if (this.wallet.sdk === 'ERC20' && fee) {
        const parentSymbol = this.supportedCoins.find((coin) => {
          return coin.name === this.wallet.parentName;
        }).symbol;

        const parentPrice = this.$store.getters['entities/latestPrice/find'](`${parentSymbol}_${this.selectedCurrency.code}`).data.PRICE;
        formattedAmount = new AmountFormatter({
          amount,
          rate: parentPrice,
          format: '0,0[.]00',
          coin: this.wallet.parentName,
          prependPlusOrMinus: false,
          currency: this.selectedCurrency,
          toCurrency: true,
          withCurrencySymbol: true,
        });
      }

      return formattedAmount.getFormatted();
    },
    goBack() {
      this.confirmSendModalOpened = false;
    },
    confirm() {
      this.loading = true;
      setTimeout(async () => {
        try {
          await this.broadcastTx();
        } catch (err) {
          console.log('cought', err);
          this.errorHandler(err);
        }
      }, this.delay.short);
    },
    completeTransaction() {
      // @todo, don't use app global, should work
      /* eslint-disable-next-line */
      app.$root.$emit('sendSuccessModalOpened', true, this.txData);

      setTimeout(() => {
        this.loading = false;
        this.confirmSendModalOpened = false;
      }, this.delay.short);
    },
  },
};
</script>

<style scoped>
.small-text {
  font-size: 0.8rem;
}

.sending-spinner-overlay {
  position: absolute;
  height: 100vh;
  width: 100vw;
  z-index: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background: rgba(255,255,255,0.9);
  opacity: 0;
  transition: all ease-in-out 250ms;
}

.sending-spinner-overlay.active{
  opacity: 1;
  z-index: 2;
}

.break {
  word-break: break-all;
}
</style>
