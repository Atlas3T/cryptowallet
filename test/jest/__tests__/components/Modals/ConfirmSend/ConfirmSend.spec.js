/* eslint-disable no-magic-numbers */
import { mount } from '@vue/test-utils';
import ConfirmSend from '@/components/Modals/ConfirmSend/ConfirmSendContent.vue';
import ConfirmSendModal from '@/components/Modals/ConfirmSend';
import { localVue, i18n, createRouter } from '@/helpers/SetupLocalVue';
import { createMocks as createStoreMocks } from '@/store/__mocks__/store.js';
import Wallet from '@/store/wallet/entities/wallet';
import Coin from '@/store/wallet/entities/coin';
import Address from '@/store/wallet/entities/address';
import Utxo from '@/store/wallet/entities/utxo';
import LatestPrice from '@/store/latestPrice';
import Tx from '@/store/wallet/entities/tx';

const walletData = JSON.parse('{"$id":3,"id":3,"account_id":1,"name":"Ethereum","displayName":"Ethereum","symbol":"ETH","sdk":"Ethereum","network":"ETHEREUM","internalChainAddressIndex":0,"externalChainAddressIndex":1,"externalAddress":null,"confirmedBalance":10,"unconfirmedBalance":0,"imported":true,"enabled":true,"lastBlockHeight":0,"parentName":"","parentSdk":"","contractAddress":"","decimals":"","hdWallet":{"bip":49,"ext":{"xpriv":"xprv9zZZrtcYB6bg2sxhoVYYKAM8duZHP1MwV1SYNVLStU6qpY6NCKfuTCeeHe1MsBeYDt7XipqZYtshoGC9ghNZ6EVwugKxUPM7m2tFmExu2mt","xpub":"xpub6DYvGQ9S1U9yFN3AuX5YgJHsBwPmnU5nrEN9Ask4SodphLRWjrz9zzy88v6HFPGJuDoMg7hhEG2yFFhU7Djqx3AoE57hEL748iFp26JLXtC"},"int":{"xpriv":"xprv9zZZrtcYB6bg6LXLr8kweVuqCWUTc5dnFDJ526bEyz5pLHdfHRx2D1UPLE8bZy6hK8f1YpXQHeGF2FWEds5S1MSbTH2boaQRNUoz6vy4MRq","xpub":"xpub6DYvGQ9S1U9yJpboxAHx1drZkYJx1YMdcSDfpUzrYKcoD5xopyGGkonsBXbmCnHteYYMcjgWRU7EyByaGf8Mav4y5BTkYmC2TD2JVPmkJxs"},"type":1,"network":{"name":"BITCOIN_TESTNET","type":"testnet","bip":1,"segwit":true,"discovery":"http://92.207.178.198:3001/api","broadcastUrl":"https://chain.so/api/v2/send_tx/BTCTEST","feeApi":"https://api.blockcypher.com/v1/btc/main","connect":{"messagePrefix":"Bitcoin Signed Message:","bech32":"tb","bip32":{"public":70617039,"private":70615956},"pubKeyHash":111,"scriptHash":196,"wif":239}}},"erc20Wallet":""}');
const erc20WalletData = JSON.parse('{"$id":4,"id":4,"account_id":1,"name":"Catalyst","displayName":"Catalyst","symbol":"CAT","sdk":"ERC20","network":"ETHEREUM","internalChainAddressIndex":0,"externalChainAddressIndex":1,"externalAddress":null,"confirmedBalance":0,"unconfirmedBalance":0,"imported":true,"enabled":true,"lastBlockHeight":0,"parentName":"Ethereum","parentSdk":"Ethereum","contractAddress":"123","decimals":"3","hdWallet":{"bip":49,"ext":{"xpriv":"xprv9zZZrtcYB6bg2sxhoVYYKAM8duZHP1MwV1SYNVLStU6qpY6NCKfuTCeeHe1MsBeYDt7XipqZYtshoGC9ghNZ6EVwugKxUPM7m2tFmExu2mt","xpub":"xpub6DYvGQ9S1U9yFN3AuX5YgJHsBwPmnU5nrEN9Ask4SodphLRWjrz9zzy88v6HFPGJuDoMg7hhEG2yFFhU7Djqx3AoE57hEL748iFp26JLXtC"},"int":{"xpriv":"xprv9zZZrtcYB6bg6LXLr8kweVuqCWUTc5dnFDJ526bEyz5pLHdfHRx2D1UPLE8bZy6hK8f1YpXQHeGF2FWEds5S1MSbTH2boaQRNUoz6vy4MRq","xpub":"xpub6DYvGQ9S1U9yJpboxAHx1drZkYJx1YMdcSDfpUzrYKcoD5xopyGGkonsBXbmCnHteYYMcjgWRU7EyByaGf8Mav4y5BTkYmC2TD2JVPmkJxs"},"type":1,"network":{"name":"BITCOIN_TESTNET","type":"testnet","bip":1,"segwit":true,"discovery":"http://92.207.178.198:3001/api","broadcastUrl":"https://chain.so/api/v2/send_tx/BTCTEST","feeApi":"https://api.blockcypher.com/v1/btc/main","connect":{"messagePrefix":"Bitcoin Signed Message:","bech32":"tb","bip32":{"public":70617039,"private":70615956},"pubKeyHash":111,"scriptHash":196,"wif":239}}},"erc20Wallet":""}');
const bitcoinWalletData = JSON.parse('{"$id":5,"id":5,"account_id":1,"name":"Bitcoin","displayName":"Bitcoin","symbol":"BTC","sdk":"Bitcoin","network":"BITCOIN_TESTNET","internalChainAddressIndex":0,"externalChainAddressIndex":1,"externalAddress":null,"confirmedBalance":-10,"unconfirmedBalance":-10,"imported":true,"enabled":true,"lastBlockHeight":0,"parentName":"","parentSdk":"","contractAddress":"","decimals":"","hdWallet":{"bip":49,"ext":{"xpriv":"xprv9zZZrtcYB6bg2sxhoVYYKAM8duZHP1MwV1SYNVLStU6qpY6NCKfuTCeeHe1MsBeYDt7XipqZYtshoGC9ghNZ6EVwugKxUPM7m2tFmExu2mt","xpub":"xpub6DYvGQ9S1U9yFN3AuX5YgJHsBwPmnU5nrEN9Ask4SodphLRWjrz9zzy88v6HFPGJuDoMg7hhEG2yFFhU7Djqx3AoE57hEL748iFp26JLXtC"},"int":{"xpriv":"xprv9zZZrtcYB6bg6LXLr8kweVuqCWUTc5dnFDJ526bEyz5pLHdfHRx2D1UPLE8bZy6hK8f1YpXQHeGF2FWEds5S1MSbTH2boaQRNUoz6vy4MRq","xpub":"xpub6DYvGQ9S1U9yJpboxAHx1drZkYJx1YMdcSDfpUzrYKcoD5xopyGGkonsBXbmCnHteYYMcjgWRU7EyByaGf8Mav4y5BTkYmC2TD2JVPmkJxs"},"type":1,"network":{"name":"BITCOIN_TESTNET","type":"testnet","bip":1,"segwit":true,"discovery":"http://92.207.178.198:3001/api","broadcastUrl":"https://chain.so/api/v2/send_tx/BTCTEST","feeApi":"https://api.blockcypher.com/v1/btc/main","connect":{"messagePrefix":"Bitcoin Signed Message:","bech32":"tb","bip32":{"public":70617039,"private":70615956},"pubKeyHash":111,"scriptHash":196,"wif":239}}},"erc20Wallet":""}');

const coinData = JSON.parse('{"$id":"Ethereum","name":"Ethereum","displayName":"Ethereum","minConfirmations":11,"sdk":"Ethereum","symbol":"ETH","network":"ETHEREUM_ROPSTEN","denomination":"0.000000000","parentName":"","parentSdk":"","contractAddress":"","decimals":""}');
const erc20CoinData = JSON.parse('{"$id":"Catalyst","name":"Catalyst","displayName":"Catalyst","minConfirmations":11,"sdk":"ERC20","symbol":"CAT","network":"ETHEREUM_ROPSTEN","denomination":"0.000","parentName":"Ethereum","parentSdk":"Ethereum","contractAddress":"123","decimals":"4"}');
const bitcoinCoinData = JSON.parse('{"$id":"Bitcoin","name":"Bitcoin","displayName":"Bitcoin","minConfirmations":6,"sdk":"Bitcoin","symbol":"BTC","network":"BITCOIN_TESTNET","denomination":"0.00000000","parentName":"","parentSdk":"","contractAddress":"","decimals":""}');

const addressData = JSON.parse('{"$id":1,"id":1,"wallet_id":3,"account_id":1,"address":"2MwSB1utt5aMRp8tY92wNjBpb96UfpDKHX7","index":0,"chain":"external","used":false}');
const utxoData = JSON.parse('{"$id":1,"id":1,"account_id":1,"wallet_id":5,"pending":false,"address":"2MwSB1utt5aMRp8tY92wNjBpb96UfpDKHX7","amount":10,"scriptPubKey":"a9142df2990ee914d0a0c0dc8ed92abe91642d7a415b87","txid":"9e792178e63be3d05b7f03f822c060909bd9fa5c451ce4ab11c1827108c5fb6d","value":271750,"vout":0}');
const latestPriceData = JSON.parse('{"$id":"ETH_GBP","coin":"ETH","currency":"GBP","updated":1554305869988,"data":{"VOLUME24HOURTO":10147125.523436274,"PRICE":3818.12,"CHANGEPCT24HOUR":4.734003741558176}}');

const customStore = {
  state: {
    modals: {
      sendConfirmTxData: {
        transaction: {
          receiver: ['123'],
          value: 1,
          fee: 0.01,
        },
        utxo: [utxoData],
        changeAddresses: ['789'],
      },
    },
  },
};

const coinSDKSMock = {
  Bitcoin: {
    broadcastTx: jest.fn().mockImplementation(() => {
      return {
        result: true,
      };
    }),
  },
  Ethereum: {
    broadcastTx: jest.fn().mockImplementation(() => {
      return {
        result: true,
      };
    }),
  },
  ERC20: {
    broadcastTx: jest.fn(),
  },
};

const backEndServiceMock = {
  async getTransactionFee() {
    return JSON.parse('{"data":{"code":"ETH","timestamp":1554218701,"data":{"high":11281690855,"medium":11281690855,"low":11281690855}},"status":200,"statusText":"OK","headers":{"new_refresh_token":"eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkZXZpY2VJZEhhc2giOiIwLjkwNzI4MTQzMzQ1Nzk0MzMiLCJpYXQiOjE1NTQyMTg4NDMsImV4cCI6MTU1Njg5NzI0M30.S1NMyIjT0N8_rVm8xhd5Tb2YuLiU74gRDgqN6Rwg7cY","content-type":"application/json; charset=utf-8"},"config":{"transformRequest":{},"transformResponse":{},"timeout":0,"xsrfCookieName":"XSRF-TOKEN","xsrfHeaderName":"X-XSRF-TOKEN","maxContentLength":-1,"headers":{"Accept":"application/json, text/plain, */*","Authorization":"Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJkZXZpY2VJZEhhc2giOiIwLjkwNzI4MTQzMzQ1Nzk0MzMiLCJpYXQiOjE1NTQyMTc1OTMsImV4cCI6MTU1NDIyMTE5M30.CSun95n5-AIxS4r_FnIWz0Y7mumes4tN9bEW7o-OLqs"},"method":"get","url":"http://92.207.178.198:6001/fee-estimate/BTC"},"request":{"__rollbar_xhr":{"method":"GET","url":"http://92.207.178.198:6001/fee-estimate/BTC","status_code":200,"start_time_ms":1554218843339,"end_time_ms":1554218843367,"subtype":"xhr"},"__rollbar_event":{"level":"info","type":"network","timestamp_ms":1554218843366,"body":{"method":"GET","url":"http://92.207.178.198:6001/fee-estimate/BTC","status_code":200,"start_time_ms":1554218843339,"end_time_ms":1554218843367,"subtype":"xhr"},"source":"client"}}}');
  },
  loadPriceFeed: jest.fn(),
};

describe('ConfirmSend component', () => {
  let wrapper;
  let router;
  let storeMocks;

  const defaultProps = {};

  function wrapperInit(options) {
    return mount(ConfirmSend, options);
  }

  function storeInit(custom, propsData, id = 3) {
    storeMocks = createStoreMocks(custom);
    Wallet.insert({ data: [walletData, bitcoinWalletData, erc20WalletData] });
    Coin.insert({ data: [coinData, bitcoinCoinData, erc20CoinData] });
    Address.insert({ data: addressData });
    Utxo.insert({ data: utxoData });
    LatestPrice.insert({ data: latestPriceData });
    router = createRouter(storeMocks.store);
    router.push({ path: `/wallet/single/send/${id}` });
    wrapper = wrapperInit({
      i18n,
      router,
      localVue,
      propsData,
      store: storeMocks.store,
      parentComponent: ConfirmSendModal,
      mocks: {
        coinSDKS: coinSDKSMock,
        backEndService: backEndServiceMock,
        errorHandler: jest.fn(),
      },
    });
  }

  it('renders and matches snapshot', () => {
    storeInit({}, defaultProps);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('displays an ethereum transaction', async (done) => {
    storeInit(customStore);
    setTimeout(() => {
      expect(wrapper.text()).toMatch('123');
      expect(wrapper.text()).toMatch('1 ETH');
      expect(wrapper.text()).toMatch('0.01 ETH');
      expect(wrapper.vm.to).toEqual('123');
      expect(wrapper.vm.txData.transaction.value).toEqual(1);
      expect(wrapper.vm.txData.transaction.fee).toEqual(0.01);
      done();
    }, 25);
  });

  it('displays a bitcoin transaction', async (done) => {
    storeInit(customStore, {}, 5);
    setTimeout(() => {
      expect(wrapper.text()).toMatch('123');
      expect(wrapper.text()).toMatch('1 BTC');
      expect(wrapper.text()).toMatch('0.01 BTC');
      expect(wrapper.vm.to).toEqual('123');
      expect(wrapper.vm.txData.transaction.value).toEqual(1);
      expect(wrapper.vm.txData.transaction.fee).toEqual(0.01);
      done();
    }, 100);
  });

  it('displays an erc20 transaction', async (done) => {
    customStore.state.modals.sendConfirmTxData.transaction.fee = 10000000000000;
    customStore.state.modals.sendConfirmTxData.transaction.receiver = '456';
    storeInit(customStore, {}, 4);
    setTimeout(() => {
      expect(wrapper.text()).toMatch('456');
      expect(wrapper.text()).toMatch('1 CAT');
      expect(wrapper.text()).toMatch('0.00001 ETH');
      expect(wrapper.vm.to).toEqual('456');
      expect(wrapper.vm.txData.transaction.value).toEqual(1);
      expect(wrapper.vm.txData.transaction.fee).toEqual(10000000000000);
      done();
    }, 100);
  });

  it('closes the modal when the back button is clicked', async (done) => {
    storeInit(customStore);
    router.push({ path: '/wallet/' });
    router.push({ path: '/wallet/single/send/3' });
    setTimeout(() => {
      wrapper.findAll('button').at(0).trigger('click');
      setTimeout(() => {
        expect(storeMocks.actions.setConfirmSendModalOpened).toHaveBeenCalled();
        done();
      }, 25);
    }, 25);
  });

  describe('broadcastTx', () => {
    it('adds a successful ethereum transaction to the database and opens the sendSuccess modal', async (done) => {
      customStore.state.modals.sendConfirmTxData.transaction.fee = 0.01;
      customStore.state.modals.sendConfirmTxData.transaction.receiver = '123';
      storeInit(customStore);
      wrapper.findAll('button').at(1).trigger('click');
      setTimeout(() => {
        expect(Tx.all()[0].wallet_id).toEqual(3);
        expect(storeMocks.actions.setSendSuccessModalOpened).toHaveBeenCalled();
        done();
      }, 100);
    });

    it('opens the sendFailure modal if an ethereum transaction fails', async (done) => {
      coinSDKSMock.Ethereum.broadcastTx.mockImplementationOnce(() => {
        return undefined;
      });
      storeInit(customStore);
      wrapper.findAll('button').at(1).trigger('click');
      setTimeout(() => {
        done();
      }, 100);
    });

    it('adds a successful bitcoin transaction to the database and opens the sendSuccess modal', async (done) => {
      storeInit(customStore, {}, 5);
      wrapper.findAll('button').at(1).trigger('click');
      setTimeout(() => {
        expect(Tx.all()[0].wallet_id).toEqual(5);
        expect(Address.all()[1].wallet_id).toEqual(5);
        expect(storeMocks.actions.setSendSuccessModalOpened).toHaveBeenCalled();
        done();
      }, 100);
    });

    it('opens the sendFailure modal if a bitcoin transaction fails', async (done) => {
      coinSDKSMock.Bitcoin.broadcastTx.mockImplementationOnce(() => {
        return undefined;
      });
      storeInit(customStore, {}, 5);
      wrapper.findAll('button').at(1).trigger('click');
      setTimeout(() => {
        done();
      }, 100);
    });
  });

  describe('parent modal', () => {
    it('opens and closes the modal', () => {
      wrapper.vm.$parent.$store = wrapper.vm.$store;
      wrapper.vm.$parent.sendConfirmModalOpened = false;
      expect(wrapper.vm.$parent.sendConfirmModalOpened).toBe(false);
      expect(storeMocks.actions.setConfirmSendModalOpened.mock.calls[1][1]).toBe(false);
      wrapper.vm.$parent.sendConfirmModalOpened = true;
      expect(storeMocks.actions.setConfirmSendModalOpened.mock.calls[2][1]).toBe(true);
    });
  });
});
