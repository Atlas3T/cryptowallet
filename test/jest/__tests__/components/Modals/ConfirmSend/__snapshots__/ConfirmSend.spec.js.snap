// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmSend component renders and matches snapshot 1`] = `
<div>
  <div
    class="sending-spinner-overlay"
  >
    <div
      class="lds-ring"
    >
      <div />
       
      <div />
       
      <div />
       
      <div />
    </div>
  </div>
   
  <div
    class="header-section"
  >
    <div
      class="header-back-button-wrapper"
    >
      <button
        class="q-btn inline q-btn-item non-selectable icon-btn back-arrow-btn q-btn--flat q-btn--rectangle q-focusable q-hoverable"
        style="font-size: 20px;"
        tabindex="0"
        type="button"
      >
        <div
          class="q-focus-helper"
          tabindex="-1"
        />
        <div
          class="q-btn__content text-center col items-center q-anchor--skip justify-center row"
        >
          <i
            aria-hidden="true"
            class="material-icons q-icon"
          >
            arrow_back
          </i>
        </div>
      </button>
    </div>
     
    <h1
      class="header-h1"
    >
      
      Confirm
    
    </h1>
  </div>
   
  <!---->
</div>
`;
